combined-docker/gdp/apps/gdp-name-add.8
.Dd August 2, 2015
.Dt GDP-NAME-ADD 8
.Os Swarm-GDP
.Sh NAME
.Nm gdp-name-add
.Nd add entry to human-oriented to GDP-oriented name directory
.Sh SYNOPSIS
.Nm gdp-name-add
.Op Fl d
.Op Fl D Ar debug-spec
.Op Fl H Ar database-host-name
.Op Fl p Ar database-password
.Op Fl P Ar database-password-file
.Op Fl q
.Op Fl u Ar database-user-name
.Ar human-name
.Op Ar gdp-name
.Sh DESCRIPTION
.Nm
adds an entry to the mapping database for
.Ar human-name
to
.Ar gdp-name
or deletes an entry for a particular
.Ar human-name .
GDPnames are represented as base64-encoded 256-bit tokens;
human names are arbitrary user-supplied text,
typically formatted as a reversed DNS-style name,
e.g.,
.Dq edu.berkeley.eecs.eric.test.01 .
.Pp
The password for the
.Ar database-user-name
is read from the keyboard.
This does
.Em not
use the
.Va swarm.gdp.hongdb.passwd
administrative parameter since that is for the account that reads the database,
not the account that writes the database.
.Pp
This is strictly for debugging and administration;
the log creation service should do this for ordinary users.
.Ss "Warning"
This implementation is likely to change or go away entirely in the future,
and the command will probably be subsumed into a swarm service.
.Sh OPTIONS
.Bl -tag
.
.It Fl d
Delete rather than add a directory entry.
If specified, the
.Ar gdp-name
argument must not be specified.
If not specified, the
.Ar gdp-name
argument must be specified.
.
.It Fl D Ar debug-spec
Turns on debugging flags using the libep-style format.
Useful only with the code in hand.
.
.It Fl H Ar database-host-name
The DNS name of the host that maintains the database.
This overrides the
.Va swarm.gdp.hongdb.host
runtime administrative parameter.
.
.It Fl p Ar database-password
The password for the database account used to update the database.
If this is not set, the
.Fl P
is is consulted.
.
.It Fl P Ar database-password-file
The name of a file containing the database password.
This is more secure than using
.Fl p
since the password will not show up in a
.Xr ps 1
output.
If not specified and
.Fl d
is not set,
defaults to the value of the
.Va swarm.gdp.creation-service.passwd-file
administrative parameter,
which in turn defaults to
.Pa /etc/gdp/creation_service_pw.txt .
If
.Fl d
is set, defaults to the value of the
.Va swarm.gdp.hongd.admin.passwd-file
administrative parameter, which in turn defaults to
.Pa /etc/gdp/hongd_admin_pw.txt .
If these files cannot be read,
the command invoker will be prompted for a password.
.
.It Fl u Ar database-user-name
The name of the database user to update the database.
This must have INSERT (if
.Fl d
is not specified or
DELETE
(if
.Fl d
is specified)
permission on the mapping table.
Defaults to the value of the
.Va swarm.gdp.creation-service.user
or
.Va swarm.gdp.hongd.admin.user
administrative parameters depending on whether
.Fl d
is specified.
Note that this are different users from that specified in the
.Va swarm.gdp.hongdb.user
parameter.
.El
.Sh EXIT STATUS
.Bl -tag
.
.It Li EX_NOINPUT
If
.Fl d
is given but the requested record does not exist,
this status is returned.
.
.It Li EX_UNAVAILABLE
The database cannot be accessed.
.
.It Li EX_USAGE
Command line parameters are incorrect.
.
.El
.Sh ADMINISTRATIVE PARAMETERS
.Bl -tag
.
.It swarm.gdp.creation-service.user
The name of the database user with permissions to update the
human-oriented name to GDPname database.
Defaults to
.Li gdp_creation_service .
.
.It swarm.gdp.creation-service.passwd-file
The name of the default password file.
This can be overridden using the
.Fl P
flag.
Defaults to
.Pa /etc/gdp/creation_service_pw.txt .
.
.It swarm.gdp.hongd.admin.user
The name of the database user with permissions to administer the
human-oriented to GDPname database;
used when
.Fl d
is active.
Defaults to
.Li hongd_admin .
.
.It swarm.gdp.hongd.admin.passwd-file
The name of the file containing the default administrative password.
Defaults to
.Pa /etc/gdp/hongd_admin_pw.txt .
.
.El
.Pp
See
.Xr gdp 7
for documentation for
.Va swarm.gdp.hongdb.host ,
.Va swarm.gdp.hongdb.database ,
and
.Va swarm.gdp.hongdb.table .
.\".Sh ENVIRONMENT
.\".Sh FILES
.Sh SEE ALSO
.Xr gdp 7
.\".Sh EXAMPLES
.Sh BUGS
Should have some easy but reasonably secure way of reading the password
for the update user without resorting to the keyboard.


combined-docker/gdp/apps/gdp-name-add.c
/* vim: set ai sw=4 sts=4 ts=4 : */

/*
**  GDP-NAME-ADD --- add a name to the external to internal name database.
**
**	----- BEGIN LICENSE BLOCK -----
**	Applications for the Global Data Plane
**	From the Ubiquitous Swarm Lab, 490 Cory Hall, U.C. Berkeley.
**
**	Copyright (c) 2015-2019, Regents of the University of California.
**	All rights reserved.
**
**	Permission is hereby granted, without written agreement and without
**	license or royalty fees, to use, copy, modify, and distribute this
**	software and its documentation for any purpose, provided that the above
**	copyright notice and the following two paragraphs appear in all copies
**	of this software.
**
**	IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
**	SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST
**	PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION,
**	EVEN IF REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
**
**	REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
**	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
**	FOR A PARTICULAR PURPOSE. THE SOFTWARE AND ACCOMPANYING DOCUMENTATION,
**	IF ANY, PROVIDED HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO
**	OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS,
**	OR MODIFICATIONS.
**	----- END LICENSE BLOCK -----
*/

#include <ep/ep.h>
#include <ep/ep_app.h>
#include <ep/ep_dbg.h>
#include <gdp/gdp.h>
#include "../gdp/gdp_hongd.h"	// cheat

#include <mysql.h>

#include <string.h>
#include <sysexits.h>


static EP_DBG	Dbg = EP_DBG_INIT("gdp-name-add", "Update name database app");

#define MILLISECONDS		* INT64_C(1000000)

static MYSQL		*NameDb;			// name mapping database

/*
**  This is essentially a copy of the standard startup routine but
**  with different database credentials.
*/

EP_STAT
name_init(struct hongdargs *ha)
{
	EP_STAT estat = EP_STAT_OK;
	bool i_own_passwd = false;

	// open a connection to the external => internal mapping database
	if (ha->dbhost == NULL)
		ha->dbhost = ep_adm_getstrparam("swarm.gdp.hongdb.host",
								GDP_DEFAULT_HONGD_HOST);
	if (ha->dbhost == NULL)
	{
		ep_app_error("No database name available; set swarm.gdp.hongdb.host");
		estat = GDP_STAT_NAK_SVCUNAVAIL;
		goto fail0;
	}

	// read database password if we don't have one already
	if (ha->dbpasswd == NULL)
	{
		char prompt_buf[100];
		snprintf(prompt_buf, sizeof prompt_buf, "Password for %s: ", ha->dbuser);
		ha->dbpasswd = getpass(prompt_buf);
		i_own_passwd = true;
	}

	// attempt the actual connect and authentication
	_gdp_name_init(ha);

fail0:
	// make sure password isn't visible to intruders
	if (ha->dbpasswd != NULL && i_own_passwd)
		memset(ha->dbpasswd, 0, strlen(ha->dbpasswd));
	return estat;
}


int
do_insert(const char *hname, const gdp_name_t gname)
{
	int l = strlen(hname);
	char xescaped[2 * l + 1];
	char gescaped[2 * sizeof (gdp_name_t) + 1];
	char qbuf[1024];

	if (ep_dbg_test(Dbg, 9))
	{
		gdp_pname_t pname;

		ep_dbg_printf("adding %s -> %s\n",
				hname, gdp_printable_name(gname, pname));
	}

	mysql_real_escape_string(NameDb, xescaped, hname, l);
	mysql_real_escape_string(NameDb, gescaped, (const char *) gname,
					sizeof (gdp_name_t));
	const char *q =
			"INSERT INTO human_to_gdp (hname, gname)\n"
			"        VALUES('%s', '%s')";
	snprintf(qbuf, sizeof qbuf, q, xescaped, gescaped);
	ep_dbg_cprintf(Dbg, 28, "    %s\n", qbuf);
	return mysql_query(NameDb, qbuf);
}


int
do_delete(const char *hname)
{
	int l = strlen(hname);
	char xescaped[2 * l + 1];
	char qbuf[1024];

	ep_dbg_cprintf(Dbg, 9, "deleting %s\n", hname);

	mysql_real_escape_string(NameDb, xescaped, hname, l);
	const char *q =
			"DELETE FROM human_to_gdp WHERE hname = '%s'\n";
	snprintf(qbuf, sizeof qbuf, q, xescaped);
	ep_dbg_cprintf(Dbg, 28, "    %s\n", qbuf);
	return mysql_query(NameDb, qbuf);
}


void
usage(void)
{
	fprintf(stderr,
			"Usage: %s [-d] [-D dbgspec] [-H db_host] [-p passwd] [-P pw_file]\n"
			"        [-q] [-u db_user] human_name gdp_name\n"
			"    -d  delete (instead of add)\n"
			"    -D  set debugging flags\n"
			"    -H  database host name (default: %s)\n"
			"    -p  database password\n"
			"    -P  name of file containing password\n"
			"    -q  suppress output (exit status only)\n"
			"    -u  database creation service user name (default: %s)\n"
			"human_name is the human-oriented log name\n"
			"gdp_name is the base-64 encoded 256-bit internal name\n",
			ep_app_getprogname(),
			ep_adm_getstrparam("swarm.gdp.hongdb.host",
								GDP_DEFAULT_HONGD_HOST),
			ep_adm_getstrparam("swarm.gdp.creation-service.user",
								"gdp_creation_service"));
	exit(EX_USAGE);
}


int
main(int argc, char **argv)
{
	const char *hname;				// external name, human-oriented text
	const char *pname;				// internal name, base64-encoded
	gdp_name_t gname;				// internal name, binary
	const char *db_passwd_file = NULL;	// file containing associated password
	bool quiet = false;
	bool delete = false;
	int opt;
	EP_STAT estat = EP_STAT_OK;
	int xstat = EX_OK;				// exit status
	bool show_usage = false;
	const char *phase;
	struct hongdargs *ha = ep_mem_zalloc(sizeof *ha);

	// collect command-line arguments
	while ((opt = getopt(argc, argv, "dD:H:p:P:qu:")) > 0)
	{
		switch (opt)
		{
		 case 'd':
			delete = true;
			break;

		 case 'D':
			ep_dbg_set(optarg);
			break;

		 case 'H':
			ha->dbhost = optarg;
			break;

		 case 'p':
			ha->dbpasswd = optarg;
			break;

		 case 'P':
			db_passwd_file = optarg;
			break;

		 case 'q':
			quiet = true;
			break;

		 case 'u':
			ha->dbuser = optarg;
			break;

		 default:
			show_usage = true;
			break;
		}
	}
	argc -= optind;
	argv += optind;

	if (show_usage || argc-- != (delete ? 1 : 2))
		usage();

	hname = *argv++;
	if (!delete)
		pname = *argv++;

	// initialize the GDP library
	phase = "initialization";
	estat = gdp_init_phase_0(NULL, 0);
	if (!EP_STAT_ISOK(estat))
	{
		if (!quiet)
			ep_app_error("GDP Library Initialization failed");
		goto fail0;
	}

	// figure out appropriate defaults
	if (delete)
	{
		if (ha->dbuser == NULL)
			ha->dbuser = ep_adm_getstrparam("swarm.gdp.hongd.admin.user",
										"hongd_admin");
		if (db_passwd_file == NULL)
			db_passwd_file = ep_adm_getstrparam("swarm.gdp.hongd.admin.passwd-file",
										"/etc/gdp/hongd_admin_pw.txt");
	}
	else
	{
		if (ha->dbuser == NULL)
			ha->dbuser = ep_adm_getstrparam("swarm.gdp.creation-service.user",
										"gdp_creation_service");
		if (db_passwd_file == NULL)
			db_passwd_file = ep_adm_getstrparam(
								"swarm.gdp.creation-service.passwd-file",
								"/etc/gdp/creation_service_pw.txt");
	}
	if (ha->dbpasswd == NULL && db_passwd_file != NULL)
	{
		FILE *pw_fp = fopen(db_passwd_file, "r");
		if (pw_fp != NULL)
		{
			char pbuf[128];

			if (fgets(pbuf, sizeof pbuf, pw_fp) != NULL && pbuf[0] != '\0')
			{
				char *p = strchr(pbuf, '\n');
				if (p != NULL)
					*p = '\0';
				ha->dbpasswd = strdup(pbuf);
			}
			fclose(pw_fp);
		}
		else if (!quiet)
			ep_app_warn("Cannot read %s", db_passwd_file);
	}
	if (ha->dbpasswd == NULL && !quiet)
		ep_app_warn("No password for %s", ha->dbuser);

	NameDb = _gdp_hongd_conn_open(ha);

	if (!quiet)
		ep_app_info("Using HONGD server %s", ha->dbhost);

	// translate GDP name from base64 to binary
	if (!delete)
	{
		phase = "name translate";
		estat = gdp_internal_name(pname, gname);
		if (!EP_STAT_ISOK(estat))
		{
			if (!quiet)
			{
				ep_app_message(estat, "cannot decode %s", pname);
				fprintf(stderr, "\t(Should be base64-encoded log name)\n");
			}
			xstat = EX_DATAERR;
			goto fail0;
		}
	}

	// figure out ultimate human name (perhaps using $GDP_NAME_ROOT)
	const char *root = getenv("GDP_NAME_ROOT");
	ep_dbg_cprintf(Dbg, 10, "GDP_NAME_ROOT=%s\n", root);
	if (root != NULL && root[0] != '\0' && strchr(hname, '.') == NULL)
	{
		int plen = strlen(root) + strlen(hname) + 2;
		char *p = ep_mem_malloc(plen);
		snprintf(p, plen, "%s.%s", root, hname);
		hname = p;
	}

	if (!quiet)
	{
		if (delete)
			ep_app_info("deleting %s", hname);
		else
			ep_app_info("adding %s => %s", hname, pname);
	}

	// open database connection
	phase = "database open";
	estat = name_init(ha);
	if (!EP_STAT_ISOK(estat))
	{
		if (!quiet)
			ep_app_message(estat, "cannot initialize name database (%s)",
					mysql_error(NameDb));
		xstat = EX_UNAVAILABLE;
		goto fail0;
	}

	// now attempt the actual database update
	int istat;
	if (delete)
	{
		phase = "DELETE";
		istat = do_delete(hname);
		if (istat == 0)
		{
			my_ulonglong nrows = mysql_affected_rows(NameDb);
			if (nrows == 0)
				xstat = EX_NOINPUT;
			if (!quiet)
			{
				if (nrows == 0)
					ep_app_warn("no rows matched");
				else
					ep_app_info("%lld rows deleted", nrows);
			}
		}
	}
	else
	{
		phase = "INSERT";
		istat = do_insert(hname, gname);
	}

	// cleanup and exit
	if (istat != 0)
	{
		if (EP_STAT_ISOK(estat))
			estat = GDP_STAT_MYSQL_ERROR;
		if (!quiet)
			ep_app_error("%s failed: %s", phase, mysql_error(NameDb));
	}

fail0:
	if (NameDb != NULL)
		mysql_close(NameDb);
	NameDb = NULL;

	// OK status can have values; hide that from the user
	if (EP_STAT_ISOK(estat))
		estat = EP_STAT_OK;
	else if (xstat == EX_OK)
		xstat = EX_UNAVAILABLE;
	exit(xstat);
}
